Division by 0 error

https://jakubsokolowski.github.io/calc-web/#/tools/positional/positional-calculator?operation=division&algorithm=default&base=10&op=1323&op=0

```
main.f4e3cfff907dd1ed0158.esm.js:2 Error: The value Infinity is not in range -1 - 9
    at Function.getRepresentation (main.f4e3cfff907dd1ed0158.esm.js:2)
    at Function.getDigit (main.f4e3cfff907dd1ed0158.esm.js:2)
    at main.f4e3cfff907dd1ed0158.esm.js:2
    at rn (main.f4e3cfff907dd1ed0158.esm.js:2)
```

Need to add some kind of operands validator for current operation that given operation can define

--- a/apps/calc-web-e2e/src/integration/positional/calculator-options.spec.ts
+++ b/apps/calc-web-e2e/src/integration/positional/calculator-options.spec.ts
@@ -59,7 +59,7 @@ describe('Calculator options', () => {
     it('should prevent calculation when some operand is invalid for base', () => {
         const base = 16;
         const operands = ['FFA', 'ZXC'];
-        const message = 'Operands not valid for given base';
+        const message = 'Some operand is not valid';
 
         selectOperation('Multiplication');
         setOperationBase(base);

--- a/apps/calc-web-e2e/src/integration/positional/division/default-division.spec.ts
+++ b/apps/calc-web-e2e/src/integration/positional/division/default-division.spec.ts
@@ -1,6 +1,7 @@
 import { OperationTemplate } from '@calc/positional-calculator';
 import { AlgorithmType, MultiplicationType, OperationType } from '@calc/calc-arithmetic';
 import {
+    enterOperation, getCalculateButton,
     getDivisionResult,
     getOperationGrid,
     gridHasProperResultRow,
@@ -206,5 +207,34 @@ describe('Default Division', () => {
         operationReturnsProperResult(config, expected);
         getDivisionResult().toMatchSnapshot();
         getOperationGrid().toMatchSnapshot();
+    });
+
+    it('should divide 0 by num', () => {
+        const base = 10;
+        const config: OperationTemplate<AlgorithmType> = {
+            operands: ['0', '100'],
+            operation: OperationType.Division,
+            algorithm: MultiplicationType.Default,
+            base
+        };
+        const expected = '0';
+
+        operationReturnsProperResult(config, expected);
+        getDivisionResult().toMatchSnapshot();
+        getOperationGrid().toMatchSnapshot();
+    });
+
+    it.only('should prevent division by 0', () => {
+        const base = 10;
+        const config: OperationTemplate<AlgorithmType> = {
+            operands: ['100', '0'],
+            operation: OperationType.Division,
+            algorithm: MultiplicationType.Default,
+            base
+        };
+
+        enterOperation(config);
+        getCalculateButton().should('be.disabled');
+        cy.get('.MuiFormHelperText-root').contains('Cannot divide by 0');
     });
 });

--- a/apps/calc-web-e2e/src/support/positional-calculator.ts
+++ b/apps/calc-web-e2e/src/support/positional-calculator.ts
@@ -49,12 +49,16 @@ export const getOperationGrid = () => cy.get('#operation-grid');
 export const getOperationGridSaveButton = () => cy.getByDataTest('operation-grid-save');
 
 export const calculatePositional = (config: OperationTemplate<AlgorithmType>) => {
+    enterOperation(config);
+    getCalculateButton().click();
+};
+
+export const enterOperation = (config: OperationTemplate<AlgorithmType>) => {
     const { algorithm, base, operands, operation } = config;
     setOperationBase(base);
     selectAlgorithm(algorithm);
     selectOperation(operation);
     operands.forEach((op, idx) => addOperand(op, idx));
-    getCalculateButton().click();
 };
 
 export const operationReturnsProperResult = (config: OperationTemplate<AlgorithmType>, result: string) => {

--- a/libs/calc-arithmetic/src/lib/calc-arithmetic.ts
+++ b/libs/calc-arithmetic/src/lib/calc-arithmetic.ts
@@ -69,3 +69,4 @@ export { ComplementConversionResult } from './models/complement';
 export { multiplyWithoutExtension } from './positional/multiplication';
 export { BoothConverter } from './positional/signed-digit/booth-converter';
 export * from './positional/signed-digit/signed-digit-converter';
+export * from './positional/validators';

--- a/libs/calc-arithmetic/src/lib/models/operation-algorithm.ts
+++ b/libs/calc-arithmetic/src/lib/models/operation-algorithm.ts
@@ -1,4 +1,5 @@
 import { OperationType } from './operation';
+import { OperandValidator } from '../positional/validators';
 
 export enum AdditionType {
     Default = 'Default',
@@ -27,66 +28,8 @@ export interface OperationAlgorithm<T extends AlgorithmType = AlgorithmType> {
     type: T;
     tKey: string;
     allowedBases?: number[];
+    operandValidators?: OperandValidator[];
 }
 
 export type AlgorithmOperationMap = Record<OperationType, OperationAlgorithm[]>;
 
-export const additionAlgorithms: OperationAlgorithm<AdditionType>[] = [
-    {
-        type: AdditionType.Default,
-        tKey: 'operations.addition.default',
-    },
-];
-
-export const subtractionAlgorithms: OperationAlgorithm<SubtractionType>[] = [
-    {
-        type: SubtractionType.Default,
-        tKey: 'operations.subtraction.default',
-    },
-];
-
-export const multiplicationAlgorithms: OperationAlgorithm<
-    MultiplicationType
->[] = [
-    {
-        type: MultiplicationType.Default,
-        tKey: 'operations.multiplication.default',
-    },
-    {
-        type: MultiplicationType.WithExtension,
-        tKey: 'operations.multiplication.withExtension',
-    },
-    {
-        type: MultiplicationType.WithoutExtension,
-        tKey: 'operations.multiplication.withoutExtension',
-    },
-    {
-        type: MultiplicationType.Booth,
-        tKey: 'operations.multiplication.booth',
-        allowedBases: [2],
-    },
-    {
-        type: MultiplicationType.BoothMcSorley,
-        tKey: 'operations.multiplication.boothMcSorley',
-        allowedBases: [2],
-    },
-    {
-        type: MultiplicationType.BoothMcSorleyAlt,
-        tKey: 'operations.multiplication.boothMcSorleyAlt',
-        allowedBases: [2],
-    },
-];
-
-export const divisionAlgorithms: OperationAlgorithm<DivisionType>[] = [
-    {
-        type: DivisionType.Default,
-        tKey: 'operations.division.default',
-    },
-];
-
-export const algorithmMap: AlgorithmOperationMap = {
-    [OperationType.Addition]: additionAlgorithms,
-    [OperationType.Subtraction]: subtractionAlgorithms,
-    [OperationType.Multiplication]: multiplicationAlgorithms,
-    [OperationType.Division]: divisionAlgorithms,
-};

--- a/libs/calc-arithmetic/src/lib/models/operation.ts
+++ b/libs/calc-arithmetic/src/lib/models/operation.ts
@@ -13,29 +13,3 @@ export interface Operation {
     tKey: string;
 }
 
-export const allOperations: Operation[] = [
-    {
-        type: OperationType.Addition,
-        maxOperands: 10,
-        minOperands: 2,
-        tKey: 'operations.addition.title'
-    },
-    {
-        type: OperationType.Subtraction,
-        minOperands: 2,
-        maxOperands: 2,
-        tKey: 'operations.subtraction.title'
-    },
-    {
-        type: OperationType.Multiplication,
-        minOperands: 2,
-        maxOperands: 2,
-        tKey: 'operations.multiplication.title'
-    },
-    {
-        type: OperationType.Division,
-        minOperands: 2,
-        maxOperands: 2,
-        tKey: 'operations.division.title'
-    }
-];

--- a/libs/calc-arithmetic/src/lib/positional/validators.ts
+++ b/libs/calc-arithmetic/src/lib/positional/validators.ts
@@ -1,0 +1,15 @@
+export type TranslationErrorMessage = {
+    key: string;
+    options?: Record<string, any>;
+}
+
+export interface OperandInputValue {
+    base: number;
+    representation: string;
+    index: number;
+    totalNumOperands: number;
+}
+
+export type OperandValidator = (input: OperandInputValue) => TranslationErrorMessage | undefined;
+
+

--- a/libs/i18n/src/assets/en.json
+++ b/libs/i18n/src/assets/en.json
@@ -84,7 +84,7 @@
         "resultPossiblyWrong": "Result is possibly wrong",
         "newOperand": "New operand",
         "maxOpReached": "Maximum number of operands for operation reached",
-        "operandsNotValid": "Operands not valid for given base",
+        "operandsNotValid": "Some operand is not valid",
         "wrongOperandsNum": "Operation allows between {{minOperands}} and {{maxOperands}} operands",
         "calculate": "Calculate",
         "correction": "Add correction: ",
@@ -127,7 +127,8 @@
         },
         "division": {
             "title": "Division",
-            "default": "Default"
+            "default": "Default",
+            "divisionByZero": "Cannot divide by 0"
         }
     },
     "sandbox": {

--- a/libs/i18n/src/assets/pl.json
+++ b/libs/i18n/src/assets/pl.json
@@ -84,7 +84,7 @@
         "resultPossiblyWrong": "Wynik prawdopodobnie jest nieprawidłowy",
         "newOperand": "Nowy operand",
         "maxOpReached": "Osiągnięto max operandów dla działania",
-        "operandsNotValid": "Operandy niepoprawne dla danej podstawy",
+        "operandsNotValid": "Niektóre operandy są niepoprawne",
         "wrongOperandsNum": "Działanie wymaga od {{minOperands}} do {{maxOperands}} operandów",
         "additionDetails": "Szczegóły dodawania",
         "subtractionDetails": "Szczegóły odejmowania",
@@ -131,7 +131,8 @@
         },
         "division": {
             "title": "Dzielenie",
-            "default": "Domyślny"
+            "default": "Domyślny",
+            "divisionByZero": "Nie można podzielić przez 0"
         }
 
     },

--- a/libs/positional-calculator/src/lib/calculator-options/algorithms.ts
+++ b/libs/positional-calculator/src/lib/calculator-options/algorithms.ts
@@ -1,0 +1,65 @@
+import {
+    AdditionType,
+    AlgorithmOperationMap,
+    DivisionType,
+    MultiplicationType,
+    OperationAlgorithm,
+    OperationType,
+    SubtractionType
+} from '@calc/calc-arithmetic';
+import { isDivisorZero } from '../validators/validators';
+
+export const additionAlgorithms: OperationAlgorithm<AdditionType>[] = [
+    {
+        type: AdditionType.Default,
+        tKey: 'operations.addition.default'
+    }
+];
+export const subtractionAlgorithms: OperationAlgorithm<SubtractionType>[] = [
+    {
+        type: SubtractionType.Default,
+        tKey: 'operations.subtraction.default'
+    }
+];
+export const multiplicationAlgorithms: OperationAlgorithm<MultiplicationType>[] = [
+    {
+        type: MultiplicationType.Default,
+        tKey: 'operations.multiplication.default'
+    },
+    {
+        type: MultiplicationType.WithExtension,
+        tKey: 'operations.multiplication.withExtension'
+    },
+    {
+        type: MultiplicationType.WithoutExtension,
+        tKey: 'operations.multiplication.withoutExtension'
+    },
+    {
+        type: MultiplicationType.Booth,
+        tKey: 'operations.multiplication.booth',
+        allowedBases: [2]
+    },
+    {
+        type: MultiplicationType.BoothMcSorley,
+        tKey: 'operations.multiplication.boothMcSorley',
+        allowedBases: [2]
+    },
+    {
+        type: MultiplicationType.BoothMcSorleyAlt,
+        tKey: 'operations.multiplication.boothMcSorleyAlt',
+        allowedBases: [2]
+    }
+];
+export const divisionAlgorithms: OperationAlgorithm<DivisionType>[] = [
+    {
+        type: DivisionType.Default,
+        tKey: 'operations.division.default',
+        operandValidators: [isDivisorZero]
+    }
+];
+export const algorithmMap: AlgorithmOperationMap = {
+    [OperationType.Addition]: additionAlgorithms,
+    [OperationType.Subtraction]: subtractionAlgorithms,
+    [OperationType.Multiplication]: multiplicationAlgorithms,
+    [OperationType.Division]: divisionAlgorithms
+};

--- a/libs/positional-calculator/src/lib/calculator-options/calculator-options.spec.tsx
+++ b/libs/positional-calculator/src/lib/calculator-options/calculator-options.spec.tsx
@@ -4,8 +4,9 @@ import { CalculatorOptions } from './calculator-options';
 import { DndOperand } from '../operand-list/operand-list';
 import { tick } from '@calc/utils';
 import { OperandInput } from '../operand-input/operand-input';
-import { allOperations, multiplicationAlgorithms } from '@calc/calc-arithmetic';
 import { act } from 'react-dom/test-utils';
+import { allOperations } from './operations';
+import { multiplicationAlgorithms } from './algorithms';
 
 const history = {
     push: jest.fn(),

--- a/libs/positional-calculator/src/lib/calculator-options/calculator-options.tsx
+++ b/libs/positional-calculator/src/lib/calculator-options/calculator-options.tsx
@@ -1,11 +1,8 @@
 import React, { FC, useEffect, useState } from 'react';
 import { ExtendedSelect, FormErrors } from '@calc/common-ui';
 import {
-    algorithmMap,
-    allOperations,
     BaseDigits,
-    isValidComplementOrRepresentationStr,
-    multiplicationAlgorithms,
+    OperandInputValue,
     Operation,
     OperationAlgorithm,
     OperationType
@@ -18,6 +15,9 @@ import { makeStyles } from '@material-ui/core/styles';
 import { DndOperand, OperandList } from '../operand-list/operand-list';
 import { useUrlCalculatorOptions } from './url-calculator-options';
 import { useHistory } from 'react-router-dom';
+import { representationValidator, validateOperand } from '../validators/validators';
+import { allOperations } from './operations';
+import { algorithmMap, multiplicationAlgorithms } from './algorithms';
 
 interface FormValues {
     base: number;
@@ -133,10 +133,18 @@ export const CalculatorOptions: FC<P> = ({ onSubmit, onOperationChange, defaultO
     const form = useFormik({ initialValues, validate, onSubmit: handleSubmit });
 
     const handleOperandChange = (newOperands: DndOperand[]) => {
-        const ops: DndOperand[] = newOperands.map((op: DndOperand) => {
+        const ops: DndOperand[] = newOperands.map((op: DndOperand, index) => {
+            const input: OperandInputValue = {
+                base: form.values.base,
+                representation: op.representation,
+                index: index,
+                totalNumOperands: newOperands.length
+            };
+            const validators = [representationValidator, ...(algorithm.operandValidators || [])];
+            const err = validateOperand(validators, input);
             return {
                 ...op,
-                valid: isValidComplementOrRepresentationStr(op.representation, form.values.base)
+                valid: !err
             };
         });
         setOperands(ops);
@@ -261,6 +269,7 @@ export const CalculatorOptions: FC<P> = ({ onSubmit, onOperationChange, defaultO
             </div>
             <div className={classes.operandsBox}>
                 <OperandList
+                    validators={algorithm.operandValidators}
                     operands={operands}
                     onChange={handleOperandChange}
                     inputBase={form.values.base}

--- a/libs/positional-calculator/src/lib/calculator-options/operations.ts
+++ b/libs/positional-calculator/src/lib/calculator-options/operations.ts
@@ -1,0 +1,28 @@
+import { Operation, OperationType } from '@calc/calc-arithmetic';
+
+export const allOperations: Operation[] = [
+    {
+        type: OperationType.Addition,
+        maxOperands: 10,
+        minOperands: 2,
+        tKey: 'operations.addition.title'
+    },
+    {
+        type: OperationType.Subtraction,
+        minOperands: 2,
+        maxOperands: 2,
+        tKey: 'operations.subtraction.title'
+    },
+    {
+        type: OperationType.Multiplication,
+        minOperands: 2,
+        maxOperands: 2,
+        tKey: 'operations.multiplication.title'
+    },
+    {
+        type: OperationType.Division,
+        minOperands: 2,
+        maxOperands: 2,
+        tKey: 'operations.division.title'
+    }
+];

--- a/libs/positional-calculator/src/lib/calculator-options/url-calculator-options.ts
+++ b/libs/positional-calculator/src/lib/calculator-options/url-calculator-options.ts
@@ -1,7 +1,9 @@
-import { CalculatorOptionsValue } from './calculator-options-value';
-import { algorithmMap, allOperations, BaseDigits, isValidComplementOrRepresentationStr } from '@calc/calc-arithmetic';
+ import { CalculatorOptionsValue } from './calculator-options-value';
+import { BaseDigits, isValidComplementOrRepresentationStr } from '@calc/calc-arithmetic';
 import { DndOperand } from '../operand-list/operand-list';
 import { inRangeInclusive, useUrlParams } from '@calc/utils';
+import { allOperations } from './operations';
+import { algorithmMap } from './algorithms';
 
 export function urlParamsToCalculatorOptionsValue(params: URLSearchParams): CalculatorOptionsValue | undefined {
     const operationStr = params.get('operation');

--- a/libs/positional-calculator/src/lib/division/division-grid/division-grid.tsx
+++ b/libs/positional-calculator/src/lib/division/division-grid/division-grid.tsx
@@ -66,8 +66,6 @@ function toCellsWithFractionSeparator(digits: DivisionOperand[], fractionPointCl
 
     if(hasFractionPart) {
         const fractionPointIndex = digitCells.length -1 - (Math.abs(resultLsp));
-        console.log(digitsToStr(digits), digitCells.length, resultLsp, digits);
-
         // Usually, separators for fraction point are defined as
         // separate vertical line in grid lines but there is an
         // an issue with line definition for lines tha start at 0

--- a/libs/positional-calculator/src/lib/operand-input/operand-input.spec.tsx
+++ b/libs/positional-calculator/src/lib/operand-input/operand-input.spec.tsx
@@ -9,6 +9,7 @@ describe('OperandInput', () => {
         const container = mount(
             <OperandInput
                 base={10}
+                numOperands={1}
                 representationStr={'0.0'}
                 index={0}
                 onRemove={jest.fn()}
@@ -31,6 +32,7 @@ describe('OperandInput', () => {
         const container = mount(
             <OperandInput
                 base={10}
+                numOperands={1}
                 representationStr={'0.0'}
                 index={0}
                 onRemove={onRemove}
@@ -60,6 +62,7 @@ describe('OperandInput', () => {
         const container = mount(
             <OperandInput
                 base={10}
+                numOperands={1}
                 representationStr={'FFE.8'}
                 index={0}
                 onRemove={onRemove}
@@ -85,6 +88,7 @@ describe('OperandInput', () => {
 
         const container = mount(
             <OperandInput
+                numOperands={1}
                 base={10}
                 representationStr={'10'}
                 index={0}

--- a/libs/positional-calculator/src/lib/operand-input/operand-input.tsx
+++ b/libs/positional-calculator/src/lib/operand-input/operand-input.tsx
@@ -4,35 +4,45 @@ import CloseIcon from '@material-ui/icons/Close';
 import { useTranslation } from 'react-i18next';
 import {
     BaseDigits,
-    isValidComplementOrRepresentationStr,
+    OperandInputValue,
+    OperandValidator,
 } from '@calc/calc-arithmetic';
+import { representationValidator, validateOperand } from '../validators/validators';
 
 interface P extends ListItemProps {
     representationStr: string;
     base: number;
     index: number;
+    numOperands: number;
+    validators?: OperandValidator[];
     onRemove: (index: number) => void;
     onRepresentationChange?: (representationStr: string, index: number, valid: boolean) => void;
     dataTest?: string;
 }
 
-export const OperandInput: FC<P> = ({ representationStr, onRepresentationChange, base, index, onRemove, dataTest, ...rest }) => {
+
+export const OperandInput: FC<P> = ({ representationStr, onRepresentationChange, base, index, onRemove, dataTest, numOperands, validators = [], ...rest }) => {
     const { t } = useTranslation();
     const [representation, setRepresentation] = useState(representationStr);
     const [error, setError] = useState<string | undefined>();
 
     useEffect(() => {
-        if(!BaseDigits.isValidBase(base)) return;
-        const validateValueStr = (valueStr: string, inputBase: number): string | undefined => {
-            if (!isValidComplementOrRepresentationStr(valueStr, inputBase)) {
-                return t(
-                    'baseConverter.wrongRepresentationStr',
-                    { base: inputBase }
-                );
-            }
+        if (!BaseDigits.isValidBase(base)) return;
+        const baseValidators: OperandValidator[] = [representationValidator];
+        const input: OperandInputValue = {
+            base,
+            representation,
+            index,
+            totalNumOperands: numOperands
         };
-        const err = !!validateValueStr(representation, base);
-        setError(validateValueStr(representation, base));
+        const err = validateOperand([...baseValidators, ...validators], input);
+
+        if (err) {
+            const { key, options } = err;
+            setError(t(key, options));
+        } else {
+            setError(undefined);
+        }
         onRepresentationChange(representation, index, !err);
         // WARNING: adding all deps here MAY cause infinite loop in
         // tests for components that use this component, so if the
@@ -41,7 +51,7 @@ export const OperandInput: FC<P> = ({ representationStr, onRepresentationChange,
         // may be an anonymous function, and those cannot be compared
         // The deps will be added by automatically by lint--fix,
         // so use it with caution
-    }, [base, representation]);
+    }, [base, representation, index]);
 
 
     return (

--- a/libs/positional-calculator/src/lib/operand-list/operand-list.tsx
+++ b/libs/positional-calculator/src/lib/operand-list/operand-list.tsx
@@ -6,6 +6,7 @@ import { OperandInput } from '../operand-input/operand-input';
 import { useTranslation } from 'react-i18next';
 import AddIcon from '@material-ui/icons/Add';
 import { reorder } from '@calc/utils';
+import { OperandValidator } from '@calc/calc-arithmetic';
 
 export interface DndOperand {
     representation: string;
@@ -16,6 +17,7 @@ export interface DndOperand {
 interface P {
     inputBase: number;
     operands: DndOperand[];
+    validators?: OperandValidator[];
     onChange: (operands: DndOperand[]) => void;
     onAdd: () => void;
     canAdd: boolean;
@@ -52,7 +54,7 @@ const useStyles = makeStyles((theme: Theme) =>
 );
 
 
-export const OperandList: FC<P> = ({ inputBase, operands, onChange, onAdd, canAdd }) => {
+export const OperandList: FC<P> = ({ inputBase, operands, onChange, onAdd, canAdd, validators }) => {
     const classes = useStyles();
     const { t } = useTranslation();
 
@@ -87,6 +89,8 @@ export const OperandList: FC<P> = ({ inputBase, operands, onChange, onAdd, canAd
             <Draggable key={item.dndKey} draggableId={`${item.dndKey}`} index={index}>
                 {(provided, snapshot) => (
                     <OperandInput
+                        validators={validators}
+                        numOperands={operands.length}
                         dataTest={`operand-input-${index}`}
                         ContainerComponent="li"
                         ContainerProps={{ ref: provided.innerRef }}

--- a/libs/positional-calculator/src/lib/validators/validators.spec.ts
+++ b/libs/positional-calculator/src/lib/validators/validators.spec.ts
@@ -1,0 +1,101 @@
+import { OperandInputValue, OperandValidator, TranslationErrorMessage } from '@calc/calc-arithmetic';
+import { isDivisorZero, validateOperand } from './validators';
+
+
+describe('#validators', () => {
+    describe('#validateOperand', () => {
+        const notBinary: OperandValidator = (input) => {
+            if (input.base === 2) {
+                return {
+                    key: 'binaryBad'
+                };
+            }
+        };
+
+        const shorterThan5: OperandValidator = (input) => {
+            if (input.representation.length >= 5) {
+                return {
+                    key: 'longBad'
+                };
+            }
+        };
+
+        it('should return first error reported by validators', () => {
+            // given
+            const validators: OperandValidator[] = [
+                notBinary, shorterThan5
+            ];
+
+            const input: OperandInputValue = {
+                representation: '111000',
+                base: 2,
+                index: 0,
+                totalNumOperands: 1
+            };
+
+            // when
+            const err = validateOperand(validators, input);
+
+            // then
+            const expected: TranslationErrorMessage = {
+                key: 'binaryBad'
+            };
+            expect(err).toEqual(expected);
+        });
+
+        it('should return undefined when validators do not report any errors', () => {
+            // given
+            const validators: OperandValidator[] = [
+                notBinary, shorterThan5
+            ];
+
+            const input: OperandInputValue = {
+                representation: '10',
+                base: 10,
+                index: 0,
+                totalNumOperands: 1
+            };
+
+            // when
+            const err = validateOperand(validators, input);
+
+            // then
+            const expected = undefined;
+            expect(err).toEqual(expected);
+        });
+    });
+
+    describe('#isDivisorZero', () => {
+        it('should return error when last operand is 0', () => {
+            // given
+            const input: OperandInputValue = {
+                representation: '0',
+                base: 10,
+                index: 1,
+                totalNumOperands: 2
+            };
+
+            // when
+            const err = isDivisorZero(input);
+
+            // then
+            expect(err).toBeDefined();
+        });
+
+        it('should return undefined when last operand is not 0', () => {
+            // given
+            const input: OperandInputValue = {
+                representation: '21',
+                base: 10,
+                index: 1,
+                totalNumOperands: 2
+            };
+
+            // when
+            const err = isDivisorZero(input);
+
+            // then
+            expect(err).toBeUndefined();
+        });
+    });
+});

--- a/libs/positional-calculator/src/lib/validators/validators.ts
+++ b/libs/positional-calculator/src/lib/validators/validators.ts
@@ -1,0 +1,40 @@
+import {
+    fromStringDirect,
+    isValidComplementOrRepresentationStr,
+    OperandInputValue,
+    OperandValidator,
+    TranslationErrorMessage
+} from '@calc/calc-arithmetic';
+
+
+export const isDivisorZero: OperandValidator = (input) => {
+    const { totalNumOperands, index, representation, base } = input;
+    if (totalNumOperands !== 2) return undefined;
+    if (index !== 1) return undefined;
+    if (isValidComplementOrRepresentationStr(representation, base)) {
+        const num = fromStringDirect(representation, base).result;
+        if (num.toNumber() === 0) {
+            return {
+                key: 'operations.division.divisionByZero'
+            };
+        }
+    }
+};
+
+export const representationValidator: OperandValidator = (input: OperandInputValue) => {
+    const { representation, base } = input;
+    if (!isValidComplementOrRepresentationStr(representation, base)) {
+        return {
+            key: 'baseConverter.wrongRepresentationStr',
+            options: { base }
+        };
+    }
+};
+
+export function validateOperand(validators: OperandValidator[], input: OperandInputValue): TranslationErrorMessage | undefined {
+    for (const validator of validators) {
+        const err = validator(input);
+        if (err) return err;
+    }
+    return undefined;
+}

